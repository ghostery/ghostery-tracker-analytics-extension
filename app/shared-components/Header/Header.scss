@import '../../utils/scss/color-variables.scss';

.Header {
  padding: 10px 0 9px;
}

.Header__logo {
  padding: 0;
  text-decoration: none;

  > img {
    position: relative;
    width: 123px;
    height: 30px;
  }

  &.panel > img {
    top: 1px;
    width: 98px;
    height: 24px;
  }
}

.Header__parentTabClosedText {
  margin: 0 60px 0 0;
  font-weight: 600;
  color: $ghosty-blue;
}

.Header__freeTrialDays {
  margin: 0 10px 0 0;
  font-weight: bold;
  font-size: 14px;
  text-align: center;
  color: $info-red;

  &.panel {
    margin: 0 0 0 15px;
    font-size: 12px;
  }
}

.Header__link {
  text-decoration: underline;
  cursor: pointer;
}

.Header__exportIcon {
  margin-left: 30px;
  margin-right: 27px;
  width: 18px;
  height: 20px;
  cursor: pointer;

  &.panel {
    margin-left: 22px;
    margin-right: 19px;
  }

  &:hover > path {
    fill: $ghosty-blue;
  }

  &.disabled {
    pointer-events: none;
  }

  &.disabled > path {
    fill: $medium-grey;
  }
}

.Header__dropdownToggle {
  cursor: pointer;

  &:hover > path {
    fill: $ghosty-blue;
  }
}

.Header__dropdownMenu {
  top: -18px;
  right: -18px;
  opacity: 1;
  background-color: $opaque-background-off-white;
  border-radius: 0;
  padding: 0;
  width: 300px;
  height: 700px;
  font-size: 14px;
  text-align: center;

  // Flex properties need to be defined here since they'll interfere with the
  // closed/open state of the dropdown if defined in the class name

  &.show {
    display: flex;
    align-items: flex-start;
    justify-content: center;
  }

  &.panel {
    position: fixed;
    top: -3px;
    right: -1px;
    height: 100vh;
  }
}
