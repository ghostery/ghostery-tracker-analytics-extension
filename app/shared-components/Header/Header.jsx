import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Navbar from 'react-bootstrap/Navbar';
import ClassNames from 'classnames';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Tooltip from 'react-bootstrap/Tooltip';
import Dropdown from 'react-bootstrap/Dropdown';

import LiveFreezeToggle from '../LiveFreezeToggle';
import Settings from '../Settings';
import MoreDropdownToggle from './MoreDropdownToggle.jsx';
import './Header.scss';

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = ({ dropdownToggle: false });
  }

  toggleDropdown = (show) => {
    this.setState({ dropdownToggle: show });
  }

  _renderFreeTrialNotification() {
    const {
      messageCreators,
      panel,
      userInfo: {
        freeTrialDaysRemaining,
      },
    } = this.props;

    const freeTrialDaysClasses = ClassNames('Header__freeTrialDays', { panel });

    return (
      <p className={freeTrialDaysClasses}>
        {freeTrialDaysRemaining}
        <span>{` trial day${freeTrialDaysRemaining === 1 ? '' : 's'} left.`}</span>
        <span> </span>
        {panel && (<br />)}
        <span
          className="Header__link"
          onClick={() => messageCreators.subscribe()}
        >
          Upgrade to Plus now.
        </span>
      </p>
    );
  }

  _showDropdown() {
    const { dropdownToggle } = this.state;
    return dropdownToggle;
  }

  render() {
    const {
      children,
      panel,
      tooltipRef,
      messageCreators,
      parentTabClosed,
      userInfo: {
        freeTrial,
      },
      isPageNotScanned,
    } = this.props;

    const logoClasses = ClassNames('Header__logo',
      'd-flex',
      'align-items-center',
      { panel });
    const exportIconClasses = ClassNames('Header__exportIcon', {
      panel,
      disabled: parentTabClosed,
    });
    const dropdownMenuClasses = ClassNames('Header__dropdownMenu', { panel });

    const parentTabClosedText = 'Parent tab has been closed and data will not be updated. Please start a new session.';

    return (
      <Navbar className="Header d-flex justify-content-between">
        <div className="d-flex align-items-center">
          <Navbar.Brand className={logoClasses}>
            <img
              alt="Insights by Ghostery Icon"
              src={[chrome.extension.getURL('dist/images/logo-icons/insights-white-text-beta.svg')]}
            />
          </Navbar.Brand>
          {children}
        </div>
        {freeTrial && this._renderFreeTrialNotification()}
        <div className="d-flex align-items-center">
          {parentTabClosed && (
            <p className="Header__parentTabClosedText">{parentTabClosedText}</p>
          )}
          <LiveFreezeToggle panel={panel} disabled={parentTabClosed} />
          <OverlayTrigger
            container={tooltipRef}
            placement="bottom"
            overlay={(
              <Tooltip className="tooltipLeft">
                Export tracker list (CSV) and raw page data/settings (JSON)
              </Tooltip>
            )}
            delay={{ show: 500 }}
          >
            <svg
              className={exportIconClasses}
              xmlns="http://www.w3.org/2000/svg"
              width="18"
              height="20"
              onClick={() => {
                messageCreators.downloadDataFiles();
                messageCreators.sendMetrics({
                  type: 'export_data',
                  insightsView: panel ? '0' : '1',
                });
              }}
            >
              <path fill="#FFF" d="M9.9 3.233v9.061c0 .522-.4.944-.9.944-.497 0-.9-.423-.9-.944V3.23L7.083 4.298a.864.864 0 0 1-1.266-.006.98.98 0 0 1-.006-1.332L8.368.274A.864.864 0 0 1 8.996 0l.006.003c.229 0 .456.091.627.27l2.557 2.687a.973.973 0 0 1-.005 1.332.87.87 0 0 1-1.267.006L9.9 3.233zm-8.1 14.26h14.4V10.87c0-.52.4-.941.9-.941.497 0 .9.424.9.94v7.574c0 .26-.1.495-.262.666a.881.881 0 0 1-.637.276H.9a.875.875 0 0 1-.634-.275.97.97 0 0 1-.265-.667V10.87c0-.52.4-.941.9-.941.497 0 .9.424.9.94v6.624z" />
            </svg>
          </OverlayTrigger>
          <Dropdown
            alignRight
            onToggle={(openSettings) => {
              this.toggleDropdown(openSettings);
              if (openSettings) {
                messageCreators.sendMetrics({
                  type: 'open_settings_panel',
                  insightsView: panel ? '0' : '1',
                });
              }
            }}
            onClick={e => e.stopPropagation()}
            show={this._showDropdown()}
          >
            <Dropdown.Toggle as={MoreDropdownToggle} />
            {!isPageNotScanned && (
              <Dropdown.Menu className={dropdownMenuClasses}>
                <Settings closeDropdown={() => this.toggleDropdown(false)} panel={panel} />
              </Dropdown.Menu>
            )}
          </Dropdown>
        </div>
      </Navbar>
    );
  }
}

Header.propTypes = {
  panel: PropTypes.bool,
  children: PropTypes.element,
  tooltipRef: PropTypes.instanceOf(PropTypes.element),
  userInfo: PropTypes.shape({
    signedIn: PropTypes.bool.isRequired,
    insightsUser: PropTypes.bool.isRequired,
    freeTrial: PropTypes.bool.isRequired,
    freeTrialDaysRemaining: PropTypes.number.isRequired,
  }).isRequired,
  messageCreators: PropTypes.shape({
    downloadDataFiles: PropTypes.func.isRequired,
    sendMetrics: PropTypes.func.isRequired,
    subscribe: PropTypes.func.isRequired,
  }).isRequired,
  parentTabClosed: PropTypes.bool,
  isPageNotScanned: PropTypes.bool.isRequired,
};

Header.defaultProps = {
  panel: false,
  children: null,
  tooltipRef: null,
  parentTabClosed: false,
};

export default Header;
