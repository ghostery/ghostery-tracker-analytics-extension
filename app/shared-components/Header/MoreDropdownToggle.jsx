import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './Header.scss';

// NOTE: This toggle component must be an extension of React.Component rather than
// a functional component because the Boostrap <Dropdown /> component requires that
// custom toggle and menu components be able to accept refs

// NOTE: I imagine we will end making a custom component for the settings dropdown
// menu as well, which works in the same way as the toggle (ie. as={CustomMenu})

class CustomMoreToggle extends Component {
  handleClick = (e) => {
    e.preventDefault();

    const { onClick } = this.props;
    onClick(e);
  }

  render() {
    return (
      <svg
        xmlns="http://www.w3.org/2000/svg"
        className="Header__dropdownToggle d-inline-block align-top"
        width="10"
        height="20"
        viewBox="0 0 4 16"
        onClick={this.handleClick}
      >
        <path d="M0 0h4v4H0zm0 6h4v4H0zm0 6h4v4H0z" fill="#FFF" fillRule="evenodd" />
      </svg>
    );
  }
}

CustomMoreToggle.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default CustomMoreToggle;
