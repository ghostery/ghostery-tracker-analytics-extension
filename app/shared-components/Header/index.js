import { connect } from 'react-redux';

import Header from './Header.jsx';
import { sendMetrics, downloadDataFiles, subscribe } from '../../store/messages/messageCreators';

const mapStateToProps = state => ({
  tooltipRef: state.UserInterfaceReducer.tooltipRef,
  userInfo: state.SettingsReducer.userInfo,
  parentTabClosed: state.PageInfoReducer.parentTabClosed,
  messageCreators: { sendMetrics, downloadDataFiles, subscribe },
  isPageNotScanned: state.PageInfoReducer.isPageNotScanned,
});

export default connect(mapStateToProps)(Header);
