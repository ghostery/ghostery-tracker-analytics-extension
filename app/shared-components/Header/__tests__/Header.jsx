import React from 'react';
import { shallow } from 'enzyme';
import chrome from 'sinon-chrome';
import path from 'path';
import fakeProps from 'react-fake-props';

import Header from '../Header';

jest.mock('../../../../src/classes/Globals', () => ({
  BROWSER_INFO: { displayName: 'Chrome' },
}));

describe('app/shared-components/Header/', () => {
  describe('Snapshot tests with shallow', () => {
    test('Header component should render', () => {
      global.chrome = chrome;
      const componentPath = path.join(__dirname, '../Header.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<Header {...props} />);
      expect(component).toMatchSnapshot();
    });
  });

  describe('Functionality testing', () => {
    test('toggleDropDown should render with setState', () => {
      global.chrome = chrome;
      const componentPath = path.join(__dirname, '../Header.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<Header {...props} />);
      component.setState({ showDropdown: true });

      expect(component.find('Dropdown').length).toBe(1);
    });
  });
});
