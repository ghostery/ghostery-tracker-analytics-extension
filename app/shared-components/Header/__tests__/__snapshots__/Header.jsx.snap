// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/shared-components/Header/ Snapshot tests with shallow Header component should render 1`] = `
<Bootstrap(undefined)
  className="Header d-flex justify-content-between"
>
  <div
    className="d-flex align-items-center"
  >
    <NavbarBrand
      className="Header__logo d-flex align-items-center"
    >
      <img
        alt="Insights by Ghostery Icon"
        src={
          Array [
            undefined,
          ]
        }
      />
    </NavbarBrand>
  </div>
  <p
    className="Header__freeTrialDays"
  >
    1
    <span>
       trial day left.
    </span>
    <span>
       
    </span>
    <span
      className="Header__link"
      onClick={[Function]}
    >
      Upgrade to Plus now.
    </span>
  </p>
  <div
    className="d-flex align-items-center"
  >
    <Connect(LiveFreezeToggle)
      disabled={false}
      panel={false}
    />
    <OverlayTrigger
      container={null}
      defaultOverlayShown={false}
      delay={
        Object {
          "show": 500,
        }
      }
      overlay={
        <Tooltip
          className="tooltipLeft"
          placement="right"
        >
          Export tracker list (CSV) and raw page data/settings (JSON)
        </Tooltip>
      }
      placement="bottom"
      trigger={
        Array [
          "hover",
          "focus",
        ]
      }
    >
      <svg
        className="Header__exportIcon"
        height="20"
        onClick={[Function]}
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.9 3.233v9.061c0 .522-.4.944-.9.944-.497 0-.9-.423-.9-.944V3.23L7.083 4.298a.864.864 0 0 1-1.266-.006.98.98 0 0 1-.006-1.332L8.368.274A.864.864 0 0 1 8.996 0l.006.003c.229 0 .456.091.627.27l2.557 2.687a.973.973 0 0 1-.005 1.332.87.87 0 0 1-1.267.006L9.9 3.233zm-8.1 14.26h14.4V10.87c0-.52.4-.941.9-.941.497 0 .9.424.9.94v7.574c0 .26-.1.495-.262.666a.881.881 0 0 1-.637.276H.9a.875.875 0 0 1-.634-.275.97.97 0 0 1-.265-.667V10.87c0-.52.4-.941.9-.941.497 0 .9.424.9.94v6.624z"
          fill="#FFF"
        />
      </svg>
    </OverlayTrigger>
    <Dropdown
      alignRight={true}
      navbar={false}
      onClick={[Function]}
      onToggle={[Function]}
      show={false}
    >
      <Bootstrap(DropdownToggle)
        as={[Function]}
      />
    </Dropdown>
  </div>
</Bootstrap(undefined)>
`;
